### PR TITLE
Keras neural networks

### PR DESCRIPTION
**Pull request by blinkseb**
Note: first commit is from #51 

This PR brings to HHTools the set of tools needed to train and evaluate neural networks using Keras + Tensorflow :tada: 

  - I added a new package, `mvaTraining` (not to be confused with `mvaTraining_hh` :laughing:) which contains the tools for **training** the NN. This package must be used outside `CMSSW`, as mentioned in the README/
  - I added a tool in `histFactory_hh` to evaluate the neural network trained by Keras. The only thing to be careful about is that you **must now execute once the `add_keras_to_cmssw.sh` script** in order to register Keras and Tensorflow inside CMSSW. This must be done only once for a given release.
  - There is also the scripts I used to build tensorflow on ingrid inside the CMSSW environment. It may be useful if we want to update the Tensorflow version at some point.

TODO:
  - [x] Move Keras output from `stderr` to `stdout`

### CODE REVIEW

**Comment by OlivierBondu** [comment]
Thanks !

File: mvaTraining/common.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+            ratio = sumw_train_signal / sumw_train_background
+            self.train_background_weights *= ratio
+            print("Background training sample reweighted so that sum of event weights for signal and background match. Sum of event weight = %.4f" % (np.sum(self.train_signal_weights)))
+
+        # Create merged training and testing dataset, with targets
+        self.training_dataset = np.concatenate([self.train_signal_dataset, self.train_background_dataset])
+        self.training_weights = np.concatenate([self.train_signal_weights, self.train_background_weights])
+        self.testing_dataset = np.concatenate([self.train_signal_dataset, self.train_background_dataset])
+        self.testing_weights = np.concatenate([self.train_signal_weights, self.train_background_weights])
```
Comment (swertz) [inline mvaTraining/common.py:179]: Shouldn't it be the testing arrays here?

File: mvaTraining/common.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+        # Create one-hot vector, the target of the training
+        # A hot-vector is a N dimensional vector, where N is the number of classes
+        # Here we assume that class 0 is signal, and class 1 is background
+        # So we have [1 0] for signal and [0 1] for background
+        self.training_targets = np.array([[1, 0]] * len(self.train_signal_dataset) + [[0, 1]] * len(self.train_background_dataset))
+        self.testing_targets = np.array([[1, 0]] * len(self.train_signal_dataset) + [[0, 1]] * len(self.train_background_dataset))
```
Comment (swertz) [inline mvaTraining/common.py:186]: Same here?

File: mvaTraining/common.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+        # Shuffle everything
+        self.training_dataset, self.training_weights, self.training_targets = shuffle(self.training_dataset, self.training_weights, self.training_targets, random_state=42)
+        self.testing_dataset, self.testing_weights, self.testing_targets = shuffle(self.testing_dataset, self.testing_weights, self.testing_targets, random_state=42)
+
+    def get_training_datasets(self):
+        return self.train_signal_dataset, self.train_background_dataset
+
+    def get_testing_datasets(self):
+        return self.train_signal_dataset, self.train_background_dataset
```
Comment (swertz) [inline mvaTraining/common.py:196]: And here?

File: mvaTraining/common.py (hunk 1 above)
Comment (blinkseb) [reply mvaTraining/common.py:179]: Indeed... I have one already fixed locally, but not eveything. Thanks a lot for the review :) I'll do a PR in the near future with the non-resonant network too :)

**Pull request merged**